pkg/controller/core/workload_controller.go

```diff
@@ -103,15 +143,38 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 				"Inadmissible", fmt.Sprintf("ClusterQueue %s is inactive", cqName))
 			return ctrl.Result{}, client.IgnoreNotFound(err)
 		}
-	case admitted:
-		msg := fmt.Sprintf("Admitted by ClusterQueue %s", wl.Spec.Admission.ClusterQueue)
-		err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionTrue, "AdmissionByKueue", msg)
+	case cancellingAdmission:
+		err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionFalse,
+			"AdmissionCancelled", "Admission cancelled")
 		return ctrl.Result{}, client.IgnoreNotFound(err)
+	case admitted:
+		if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadAdmitted) {
+			return r.reconcileWithAdmittedCondition(ctx, req, &wl)
+		} else {
+			msg := fmt.Sprintf("Admitted by ClusterQueue %s", wl.Spec.Admission.ClusterQueue)
+			err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionTrue, "AdmissionByKueue", msg)
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
 	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *WorkloadReconciler) reconcileWithAdmittedCondition(ctx context.Context, req ctrl.Request, wl *kueue.Workload) (ctrl.Result, error) {
```

Comment:
Make the name specific until the name no longer makes sense.

```suggestion
func (r *WorkloadReconciler) reconcileNotReadyTimeout(ctx context.Context, req ctrl.Request, wl *kueue.Workload) (ctrl.Result, error) {
```

Reply:
Done